posthog/hogql_queries/insights/paths_query_runner.py

```diff
@@ -0,0 +1,468 @@
+from datetime import datetime, timedelta
+from math import ceil
+from re import escape
+from typing import Any, Dict, Literal, cast
+from typing import Optional
+
+from posthog.caching.insights_api import BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL, REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+from posthog.caching.utils import is_stale
+from posthog.constants import PAGEVIEW_EVENT, SCREEN_EVENT, HOGQL
+from posthog.hogql import ast
+from posthog.hogql.constants import LimitContext
+from posthog.hogql.parser import parse_select, parse_expr
+from posthog.hogql.printer import to_printed_hogql
+from posthog.hogql.property import property_to_expr
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql.timings import HogQLTimings
+from posthog.hogql_queries.query_runner import QueryRunner
+from posthog.hogql_queries.utils.query_date_range import QueryDateRange
+from posthog.models import Team
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.queries.util import correct_result_for_sampling
+from posthog.schema import (
+    HogQLQueryModifiers,
+    PathsQueryResponse,
+    PathCleaningFilter,
+    PathsFilter,
+    PathType,
+)
+from posthog.schema import PathsQuery
+
+EVENT_IN_SESSION_LIMIT_DEFAULT = 5
+SESSION_TIME_THRESHOLD_DEFAULT_MILLISECONDS = 1800000  # milliseconds to 30 minutes
+EDGE_LIMIT_DEFAULT = 50
+
+
+class PathsQueryRunner(QueryRunner):
+    query: PathsQuery
+    query_type = PathsQuery
+
+    def __init__(
+        self,
+        query: PathsQuery | Dict[str, Any],
+        team: Team,
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        limit_context: Optional[LimitContext] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, limit_context=limit_context)
+
+        if not self.query.pathsFilter:
+            self.query.pathsFilter = PathsFilter()
+
+        self.event_in_session_limit = self.query.pathsFilter.step_limit or EVENT_IN_SESSION_LIMIT_DEFAULT
+
+        self.regex_groupings: list[str] = []
+        if self.query.pathsFilter.path_groupings:
+            self.regex_groupings = [
+                escape(grouping).replace("\\*", ".*") for grouping in self.query.pathsFilter.path_groupings
+            ]
+
+    @property
+    def group_type_index(self) -> int | None:
+        return self.query.aggregation_group_type_index
+
+    def _get_event_query(self) -> list[ast.Expr]:
+        conditions: list[ast.Expr] = []
+        or_conditions: list[ast.Expr] = []
+
+        if not self.query.pathsFilter.include_event_types:
+            return []
+
+        if PathType.field_pageview in self.query.pathsFilter.include_event_types:
+            or_conditions.append(parse_expr(f"event = '{PAGEVIEW_EVENT}'"))
+
+        if PathType.field_screen in self.query.pathsFilter.include_event_types:
+            or_conditions.append(parse_expr(f"event = '{SCREEN_EVENT}'"))
```

Comment:
Let's try to avoid f-strings in HogQL land as much as possible. In this case nothing will really happen, but it sets a bad copy-pasteable example. 

```suggestion
        if PathType.field_pageview in self.query.pathsFilter.include_event_types:
            or_conditions.append(parse_expr("event = {event}", {"event": ast.Constant(value=PAGEVIEW_EVENT)}))

        if PathType.field_screen in self.query.pathsFilter.include_event_types:
            or_conditions.append(parse_expr("event = {event}", {"event": ast.Constant(value=SCREEN_EVENT)}))
```